File: 814-二叉树剪枝.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* pruneTree(TreeNode* root);
    int post_order(TreeNode* node, TreeNode* par, bool is_left);
};

int Solution::post_order(TreeNode* node, TreeNode* par, bool is_left){
    int cur_sum = 0;
    if(node != nullptr){
        cur_sum += node->val;
        cur_sum += post_order(node->left, node, true);
        cur_sum += post_order(node->right, node, false);
        if(cur_sum == 0){
            if(is_left)
                par->left = nullptr;
            else
                par->right = nullptr;
        }
    }

    return cur_sum;
}

TreeNode* Solution::pruneTree(TreeNode* root){
    if(root){
        post_order(root->left, root, true);
        post_order(root->right, root, false);
    }
    if(root == nullptr || (root->left == nullptr && root->right == nullptr && root->val == 0))
        return nullptr;
    
    return root;
}

File: 814-二叉树剪枝.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pruneTree(self, root: TreeNode) -> TreeNode:
        self.post_order(root)
        if not root or root.val == -1:
            return None
        node_stack = [root]
        while node_stack:
            node = node_stack.pop()
            if node.left and node.left.val == -1:
                node.left = None
            if node.right and node.right.val == -1:
                node.right = None
            if node.left:
                node_stack.append(node.left)
            if node.right:
                node_stack.append(node.right)

        return root

    def post_order(self, node):
        cur_sum = 0
        if node:
            cur_sum = node.val
            cur_sum += self.post_order(node.left)
            cur_sum += self.post_order(node.right)
            if cur_sum == 0:
                node.val = -1

        return cur_sum
